crypto/o_time.c

```diff
@@ -358,3 +358,19 @@ static void julian_to_date(long jd, int *y, int *m, int *d)
     *m = j + 2 - (12 * L);
     *y = 100 * (n - 49) + i + L;
 }
+
+
+int OPENSSL_timegm(struct tm *tm, time_t *t)
+{
+    /* Figures it out without mucking the environment */
+    long basetime, currtime;
+
+    basetime = date_to_julian(1970, 1, 1);
+    currtime = date_to_julian(tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday);
+
+    if (t != NULL)
+        *t = (((currtime - basetime) * 24 + tm->tm_hour) * 60 + tm->tm_min) * 60 + tm->tm_sec;
```

Comment:
OPENSSL_gmtime() is basically a wrapper around the proper selection of the POSIX functions gmtime() and gmtime_r(). I don't see why that can be considered a mistake (but I do consider the non-POSIX VMS implementation to be a mistake). The use of OPENSSL_gmtime() is a good mechanism for portability and operating system abstraction.

OpenSSL provides a way to create an ASN1_TIME from a time_t, but not the inverse. On another part of this PR, @richsalz was advocating the use of time_t as the unambiguous representation of time, but OpenSSL (as it currently stands) does not provide a mechanism to get a time_t from an ASN1_TIME.

To make an ASN1_TIME from a time_t, the standard POSIX gmtime() (or gmtime_r()) function is used (via the OPENSSL_gmtime() wrapper).

However, there is no POSIX inverse of this function that does not involve the use of timezones (something that needs to be avoided) in order to go from an ASN1_TIME to a time_t. The few ways to do that are to use the non-standard timegm(), play around with timezones and mktime(), or this function.

This function does not require the use of any system-level API to derive it's return value, which is what makes it highly portable.

By eliminating this function, the ability to have OpenSSL provide a time_t from an ASN1_TIME would also have to be eliminated, and I would argue that creating an ASN1_TIME from a time_t should be eliminated too. This basically leaves the use of **struct tm** to get and set time values that a program can use. However, **struct tm** is ambiguous, given that it has a DST field, and two values of a **struct tm** can be derived from a one value of time_t (via gmtime() and localtime).

All that being said, if this type of functionality does belongs at the application layer, then I would propose having OPENSSL_gmtime() and OPENSSL_timegm() invoke a user-provided callback to do this conversion (time_t <-> **struct tm**), leaving the _how_ to the application. However, that would likely break existing functionality and backward compatibility, so a reasonable default would need to be provided (especially for unit tests and the openssl executable).